spmd/fsdp.py

```diff
@@ -0,0 +1,325 @@
+import warnings
+import functools
+import copy
+from typing import Any, List, NamedTuple, Optional, Tuple
+
+import torch
+import torch.distributed as dist
+
+from torch.distributed.fsdp._utils import _set_fsdp_flattened
+from torch.distributed.fsdp._shard_utils import _create_chunk_sharded_tensor
+
+import torch.distributed._shard.sharding_spec as shard_spec
+from torch.distributed._shard.sharding_spec.chunk_sharding_spec import ChunkShardingSpec
+
+from torch.distributed._shard.sharded_tensor import (
+    Shard,
+    ShardedTensor,
+    ShardedTensorMetadata,
+    TensorProperties,
+)
+
+from torch.distributed._shard.sharding_spec import (
+    ShardMetadata,
+)
+
+from torch.distributed.remote_device import _remote_device
+
+from spmd.tensor import DTensor as DistributedTensor, DeviceMesh
+from spmd.tensor.placement_types import Placement
+
+__all__ = ["is_available"]
+
+class _STShardingInfo(NamedTuple):
+    """:class:`ShardedTensor` sharding information."""
+
+    sharding_spec: Optional[shard_spec.ShardingSpec]
+    global_size: Optional[torch.Size]
+    process_group: Optional[dist.ProcessGroup]
+    device_mesh: Optional[DeviceMesh]
+    placements: Optional[List[Placement]]
+
+def _get_box(tensor: DistributedTensor) -> Tuple:
+    device_mesh = tensor.device_mesh
+    assert device_mesh.ndim == 1, f"Only 1D DeviceMeshes currently handled"
+
+    placement = tensor.placements[0]
+    offsets = [0] * len(tensor.size())
+    num_chunks = device_mesh.size(dim=0)
+
+    if tensor.placements[0].is_shard():
+        shard_dim = placement.dim
+        chunk_size = tensor.size(shard_dim) // num_chunks
+        offsets[shard_dim] = chunk_size
+
+    return (torch.Size(offsets), tensor._local_tensor.size())
+
+def _get_box_for(tensor: DistributedTensor, idx: int) -> Tuple:
+    offsets, size = _get_box(tensor)
+    offsets = [val * idx for val in offsets]
+    return (torch.Size(offsets), size)
+
+def _get_local_box(tensor: DistributedTensor) -> Tuple:
+    device_mesh = tensor.device_mesh
+    return _get_box_for(tensor, device_mesh.get_coordinate_on_dim(0))
+
+def _create_shard_md_from_dt(dt: DistributedTensor, current_rank) -> ShardMetadata:
+    mesh = dt.device_mesh
+    assert mesh.ndim == 1, f"Only 1D DeviceMeshes currently handled"
+
+    offsets, sizes = _get_local_box(dt)
+    return ShardMetadata(
+        shard_offsets=list(offsets),
+        shard_sizes=list(sizes),
+        placement=f"rank:{current_rank}/{dt._local_tensor.device}"
+    )
+
+def _create_sharded_tensor_md_from_dt(dt: DistributedTensor, dt_pg) -> ShardedTensorMetadata:
```

Comment:
We cannot replace ShardedTensor with DTensor since the later cannot represent the sort of sharding we need for checkpointing.

In particular, it doesn't support uneven sharding, include ranks with missing data.

I'm not defending we stay we ShardedTensor as you might have seem by the horrible piles of hacks I had to apply for 2d parallel but we can't use DT, it has to be something else.